Call completeStream only once on failure.

gregw opened the pull request:
Fix #13470 by calling completeStream only once even when there is a failure in the channel callback.

gregw committed:
Call completeStream only once on failure.

Fix #13470 by calling completeStream only once even when there is a failure in the channel callback.

gregw commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java:
```diff
@@ -1576,7 +1576,7 @@ public void succeeded()
                         errorResponse = new ErrorResponse(request);
                     else
-                        completeStream = true;
+                        completeStream = httpChannelState._handling == null;
```
I believe this is the fix to prevent completeStream being called whilst we are still handling.  See analysis in main discussion.

gregw commented:
The issue is fundamentally that `completeStream` is being called twice.

For this to happen, we need a failure to be detected in the `ChannelCallback.succeeded()` method so that the following code isrun:
https://github.com/jetty/jetty.project/blob/8dbd9865cc7407d735084a181cc3f21205872b70/jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java#L1573-L1580

This means that  `completeStream` will be called, even though the other "legs of the 3 legged stool" are not complete - specifically we may still be inside the call to `HandlerInvoker.run()`, as in this stack for thread 258 :
```
2025-10-20 08:14:42,457 INFO  [WebServerImpl-258] trace.jetty.session.complete: complete() called on session [ManagedSession@4df349fb{id=MYSECRETSESSIONID,x=MYSECRETSESSIONID.node0,req=6,res=true}]
java.lang.Exception: complete stack
	...
	at org.eclipse.jetty.session.AbstractSessionManager.complete(AbstractSessionManager.java)
	at org.eclipse.jetty.session.AbstractSessionManager$SessionStreamWrapper.doComplete(AbstractSessionManager.java:1509)
	at org.eclipse.jetty.server.handler.ContextHandler$ScopedContext.run(ContextHandler.java:1518)
	at org.eclipse.jetty.session.AbstractSessionManager$SessionStreamWrapper.failed(AbstractSessionManager.java:1479)
	at org.eclipse.jetty.server.internal.HttpChannelState$HandlerInvoker.completeStream(HttpChannelState.java:788)
	at org.eclipse.jetty.server.internal.HttpChannelState$ChannelCallback.succeeded(HttpChannelState.java:1591)
	at org.eclipse.jetty.server.handler.gzip.GzipResponseAndCallback.succeeded(GzipResponseAndCallback.java:95)
	at org.eclipse.jetty.ee10.servlet.ServletChannel.onCompleted(ServletChannel.java:765)
	at org.eclipse.jetty.ee10.servlet.ServletChannel.handle(ServletChannel.java:429)
	at org.eclipse.jetty.ee10.servlet.ServletHandler.handle(ServletHandler.java:470)
	at org.eclipse.jetty.ee10.servlet.SessionHandler.handle(SessionHandler.java:717)
	at org.eclipse.jetty.server.handler.ContextHandler.handle(ContextHandler.java:1071)
	at org.eclipse.jetty.server.Handler$Wrapper.handle(Handler.java:740)
	at org.eclipse.jetty.rewrite.handler.RewriteHandler.handle(RewriteHandler.java:138)
	at org.eclipse.jetty.server.handler.gzip.GzipHandler.handle(GzipHandler.java:611)
	at org.eclipse.jetty.server.Handler$Sequence.handle(Handler.java:805)
	at org.eclipse.jetty.server.Server.handle(Server.java:182)
	at org.eclipse.jetty.server.internal.HttpChannelState$HandlerInvoker.run(HttpChannelState.java:677)
	at org.eclipse.jetty.util.thread.Invocable$ReadyTask.run(Invocable.java:177)
	at org.eclipse.jetty.http2.server.internal.HttpStreamOverHTTP2$1.run(HttpStreamOverHTTP2.java:144)
	...
```

We can see that there is a failure detected in `ChannelCallback.succeeded()` because `SessionStreamWrapper.failed`  is ultimately called.  This means that there must have been one of the following application errors:
 + https://github.com/jetty/jetty.project/blob/8dbd9865cc7407d735084a181cc3f21205872b70/jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java#L1543
 + https://github.com/jetty/jetty.project/blob/8dbd9865cc7407d735084a181cc3f21205872b70/jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java#L1545
 + https://github.com/jetty/jetty.project/blob/8dbd9865cc7407d735084a181cc3f21205872b70/jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java#L1563

These are all plausible application errors, especially with something like server sent events.

So once thread 258 has called `completeStream` it returns all the way out of handling and it can be seen calling `completeStream` again, in this stack trace:
```
java.lang.Exception: complete stack
	...
	at org.eclipse.jetty.session.AbstractSessionManager.complete(AbstractSessionManager.java)
	at org.eclipse.jetty.session.AbstractSessionManager$SessionStreamWrapper.doComplete(AbstractSessionManager.java:1509)
	at org.eclipse.jetty.server.handler.ContextHandler$ScopedContext.run(ContextHandler.java:1524)
	at org.eclipse.jetty.session.AbstractSessionManager$SessionStreamWrapper.failed(AbstractSessionManager.java:1479)
	at org.eclipse.jetty.server.internal.HttpChannelState$HandlerInvoker.completeStream(HttpChannelState.java:788)
	at org.eclipse.jetty.server.internal.HttpChannelState$HandlerInvoker.run(HttpChannelState.java:712)
	at org.eclipse.jetty.util.thread.Invocable$ReadyTask.run(Invocable.java:177)
	at org.eclipse.jetty.http2.server.internal.HttpStreamOverHTTP2$1.run(HttpStreamOverHTTP2.java:144)
	...
```

which is called from this code after the handler has been invoked:
https://github.com/jetty/jetty.project/blob/8dbd9865cc7407d735084a181cc3f21205872b70/jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java#L691-L713

Note that in order for this code to actually call `completeStream` then it must be `true` that `completeStream = callbackCompleted && lastStreamSendComplete`.   Note that this is normally not the case for HTTP/1 because the first call to `completeStream` would have recycled the `HttpChannelState` and the `callbackCompleted` and `lastStreamSendComplete` will both be false.   However, for H2, `HttpChannelState`s are re-used after being recycled, so another request may have come in and set the fields of the state again, so that the second call to `completeStream` incorrectly completes that new request.

Thus I believe the core fix is to not call `completeStream` whilst we are still handling.  Furthermore, if we are to ignore the last write leg of the stool, we should explicitly force `lastStreamSendComplete` to true;

Unfortunately I have been unable to produce a unit test for this, as I believe it needs precisely unlucky timing and an application error.

gregw committed:
Call completeStream only once on failure.

Fix #13470 by calling completeStream only once even when there is a failure in the channel callback.

gregw commented:
@sbordet @lorban Can you review the diagnosis that @janbartel and I have come up with.  I'm 90% sure this is it, but I cannot reproduce (any thoughts how we might be able to do that?).

gregw commented:
Note that we added this `completeStream` call in https://github.com/jetty/jetty.project/pull/9684

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java:
```diff
@@ -1574,9 +1574,14 @@ public void succeeded()
+                        // We are not committed, so we can send an error response.
                         errorResponse = new ErrorResponse(request);
-                    else
+                    else if (httpChannelState._handling == null)
```
I think this more precisely should be:


```suggestion
                    else if (httpChannelState._handled)
```

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java:
```diff
@@ -1574,9 +1574,14 @@ public void succeeded()
                         completeStream = true;
+                    else
+                        // We are committed and still handling, so let the HandlerInvoker complete, ignoring any pending reads/writes.
+                        httpChannelState._streamSendState = StreamSendState.LAST_COMPLETE;
```
Again, not sure we should ignore pending writes.

gregw replied:
Or pending reads....  it is a difficult one.   I'll at least change the code to enumerate the possible states

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java:
```diff
@@ -1574,9 +1574,14 @@ public void succeeded()
-                    else
+                    else if (httpChannelState._handling == null)
+                        // We are committed, but no longer handling, so will complete here, ignoring any pending reads/writes
                         completeStream = true;
```
I don't think this is right.

If there is a write pending, we cannot complete the stream, or we bypass one leg.
I think we should call `lockedIsLastStreamSendCompleted()` here to figure out whether that leg is done.

sbordet reviewed (changes requested):
`ChannelCallback.failed()` seems not entirely correct either.

Can we write test cases for this scenario?

gregw commented:
> `ChannelCallback.failed()` seems not entirely correct either.

@sbordet I will look....

> Can we write test cases for this scenario?

Very hard, because unless there is another thread racing the second completeStream is not called.  I'm open to suggestions.

gregw committed:
Call completeStream only once on failure.

Fix #13470 by calling completeStream only once even when there is a failure in the channel callback.

refactor success and failure into single method.

gregw committed:
Call completeStream only once on failure.

Fix #13470 by calling completeStream only once even when there is a failure in the channel callback.

refactor success and failure into single method.

Improved EventSourceServlet

gregw committed:
Call completeStream only once on failure.

Fix #13470 by calling completeStream only once even when there is a failure in the channel callback.

refactor success and failure into single method.

Improved EventSourceServlet

gregw commented:
@sbordet @lorban I'm getting concerned at the number of tests this PR is breaking in its current state.
I think we need to take time to consider the more "cleanup" changes and probably only make this in 12.1.x

So I propose that this PR should simply be https://github.com/jetty/jetty.project/pull/13814/commits/83c17183f6a100253805f435241507cf4707a9d6 for 12.0.x (perhaps with the EventSourceServlet cleanups), and then we can do a wider cleanup and refactor in 12.1.x next month.

gregw committed:
Call completeStream only once on failure.

Fix #13470 by calling completeStream only once even when there is a failure in the channel callback.

refactor success and failure into single method.

Improved EventSourceServlet

gregw committed:
Call completeStream only once on failure.

Fix #13470 by calling completeStream only once even when there is a failure in the channel callback.

refactor success and failure into single method.

Improved EventSourceServlet

gregw committed:
Call completeStream only once on failure.

Fix #13470 by calling completeStream only once even when there is a failure in the channel callback.

refactor success and failure into single method.

Improved EventSourceServlet

lorban commented:
Regarding the minimal 12.0 fix, I think `httpChannelState._handling == null` should actually be `httpChannelState._handled`

gregw commented:
@lorban this is passing tests now.... so let's go for this one?
@sbordet @lorban Review please!!

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java:
```diff
@@ -1537,115 +1554,105 @@ public void succeeded()
+                    failure = ExceptionUtil.combine(failure, stream.consumeAvailable());
+                else
+                {
+                    Throwable unconsumed = stream.consumeAvailable();
+                    if (failure != null)
+                        ExceptionUtil.addSuppressedIfNotAssociated(failure, unconsumed);
```
Just `ExceptionUtil.combine(failure, stream.consumeAvailable());` as the line above?
But then, this `else` block is identical to the `else-if` above?

gregw replied:
No this  is different.  If failure==null, then it remains null. i.e. it is not a failure to not consume.  I'll comment....

gregw replied:
Actually, I will try without the whole branch....

gregw replied:
Branch is needed

sbordet commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/ExceptionUtil.java:
```diff
@@ -191,6 +191,34 @@ public static boolean areNotAssociated(Throwable t1, Throwable t2)
+     * @param throwable The type of {@link Throwable} class to check association against. Can be null.
+     * @return true if the {@link Throwable} is associated with the provided type; otherwise, false.
+     */
+    public static boolean isAssociated(Throwable failure, Class<? extends Throwable> throwable)
```
Reads better `hasAssociated()` to me.

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java:
```diff
@@ -1537,115 +1554,105 @@ public void succeeded()
+                        else if (response.lockedIsWriting())
+                        {
+                            // We are currently writing, so let the completion of that write handle the failure
+                            httpChannelState._callbackFailure = failure;
```
Unnecessary line.

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java:
```diff
@@ -1537,115 +1554,105 @@ public void succeeded()
+                        else
+                        {
+                            // There has been no last write, but we will just fail the stream instead.
+                            completeStream = true;
```
Feels like it should be this, so the third leg is complete:


```suggestion
                            httpChannelState._streamSendState = StreamSendState.FAILED;
                            completeStream = true;
```

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java:
```diff
@@ -1537,115 +1554,105 @@ public void succeeded()
+                    {
+                        // last write is not going to happen after failure, so we can just fail anyway
+                        httpChannelState._streamSendState = StreamSendState.LAST_COMPLETE;
+                    }
```
If `lockedIsWriting()` I should do the same logic as above: steal the write callback, etc.

I would prefer this `else-if` to be written as just `else`, because the lack of a catch-all `else` branch feels like we're forgetting something: I prefer to have an explicit branch that says "here we should really do nothing".

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java:
```diff
@@ -1537,115 +1554,105 @@ public void succeeded()
                 {
+                    doLastStreamSend = httpChannelState.lockedLastStreamSend();
+                    if (doLastStreamSend)
+                        response._writeCallback = httpChannelState._handlerInvoker;
```
This feels really sneaky.

Why not just doing `response.write(true, null, httpChannelState._handlerInvoker)`?

gregw replied:
Because we have done half the write mechanism already by doing the commit at 1578 above. Calling write now will not send the headers.

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java:
```diff
@@ -619,7 +624,9 @@ public String toString()
-    private class HandlerInvoker implements Invocable.Task, Callback
+    // HandlerInvoker is used as the Response's _writeCallback when ChannelCallback is succeeded and the last send still
+    // needs to be done, i.e.: _streamSendState set to LAST_SENDING by lockedLastStreamSend().
+    private class HandlerInvoker implements Task, Callback
```
I think we should split the functionality of this class.
Leave `run()` in `HandlerInvoker`, but move the `Callback` functionality into a `LastStreamSendCallback` class.

gregw replied:
This is really not a 12.0 thing then

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java:
```diff
@@ -1537,115 +1554,105 @@ public void succeeded()
+                            // We are currently writing, so let the completion of that write handle the failure
+                            httpChannelState._callbackFailure = failure;
+                            failedCallback = response._writeCallback;
+                            response._writeCallback = httpChannelState._handlerInvoker;
```
In this way, we may wait forever for the write to complete and invoke the `_handlerInvoker`.

How about this:


```suggestion
                            Runnable task = response.lockedFailWrite(failure);
                            failedCallback = Callback.from(task, httpChannelState._handlerInvoker);
```

In 12.1.x we will leverage the new `cancelSend()` feature automatically.

gregw replied:
I'm not sure that continuing on with a pending write pointing at the HttpChannelState is a good idea.  Feels like an invitation for a race.   What is wrong waiting for the write to complete.  It won't be forever unless they have disabled idle timeout.

sbordet replied:
> I'm not sure that continuing on with a pending write pointing at the HttpChannelState is a good idea.

I'm not sure what you mean above.

My point is that failing the callback may return the buffer to the pool, while the write is still pending.
We fixed this in 12.1.x with `cancelSend()`, but we don't leverage it here, while we would if we call `lockedFailWrite()`.

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java:
```diff
@@ -93,7 +93,10 @@ private enum StreamSendState
+        LAST_COMPLETE,
+
+        /** Failing, so last send will never happen */
+        FAILED
```
This state is only read, never assigned.

gregw replied:
oops not fully pushed.

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java:
```suggestion
                completeStream = _handled;
```

gregw replied:
I tried, but we also need to complete a stream in cases where we have not started handling yet.

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java:
Retain the `InvocationType`.

gregw committed:
updates from review

gregw committed:
Merge remote-tracking branch 'origin/jetty-12.0.x' into fix/12.0.x/13470-completeStreamOnce

gregw committed:
Updates from review

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java:
```diff
@@ -1537,115 +1564,125 @@ public void succeeded()
+                        else
+                        {
+                            // There has been no last write, but we will just fail the stream instead.
+
```
Remove empty line.

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java:
```diff
@@ -1537,115 +1564,125 @@ public void succeeded()
                     else
-                        completeStream = true;
+                    {
+                        // We are still handling, sof for the most part
```
```suggestion
                        // We are still handling, so for the most part
```

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java:
```diff
@@ -1537,115 +1564,125 @@ public void succeeded()
+            if (failedCallback != null)
+                failedCallback.failed(Objects.requireNonNullElseGet(failure, IOException::new));
 
             if (errorResponse != null)
```
This should be an `if + else-if` because it is never the case that if we have a `failedCallback` we want to check the other cases below.

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java:
```diff
@@ -1537,115 +1564,125 @@ public void succeeded()
+                            failedCallback = response._writeCallback;
+                            response._writeCallback = httpChannelState._lastWriteCallback;
+                        }
+                        else if (!httpChannelState.lockedIsLastStreamSendCompleted())
+                        {
+                            // last write it is not going to happen after failure, so we can just fail anyway
+                            httpChannelState._streamSendState = StreamSendState.FAILED;
+                        }
```
I would prefer to have an explicit `else` branch, otherwise feels like we are missing something.
Even if it is `else LOG.debug("Nothing to do in this case")`.

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java:
Should not this be NON_BLOCKING?
It only completes the `HttpStream`.

gregw committed:
Updates from review

gregw committed:
updates from review
